core/station/impl/src/services/external_canister.rs

```diff
@@ -72,4 +128,357 @@ impl ExternalCanisterService {
             reason: err.to_string(),
         })
     }
+
+    /// Adds a new external canister to the system.
+    ///
+    /// Can be used to create another canister to a subnet or add an existing canister.
+    pub async fn add_external_canister(
+        &self,
+        input: CreateExternalCanisterOperationInput,
+    ) -> ServiceResult<ExternalCanister> {
+        self.check_unique_name(input.name.clone().as_str(), None)?;
+
+        let canister_id = match &input.kind {
+            CreateExternalCanisterOperationKind::CreateNew(opts) => self
+                .create_canister(opts.initial_cycles.map(|cycles| cycles as u128))
+                .await
+                .map_err(|err| ExternalCanisterError::Failed {
+                    reason: format!("failed to create external canister: {}", err),
+                })?,
+            CreateExternalCanisterOperationKind::AddExisting(opts) => {
+                EnsureExternalCanister::is_external_canister(opts.canister_id)?;
+
+                self.check_unique_canister_id(&opts.canister_id, None)?;
+
+                opts.canister_id
+            }
+        };
+
+        let external_canister = ExternalCanisterMapper::from_create_input(canister_id, input);
+
+        external_canister.validate()?;
```

Comment:
this can return an error when the canister is already created, losing it

Reply:
It could because we validate the name, description, labels, etc... but i moved the creation logic to after this validation, which means that to initially create the `ExternalCanister` struct i had to provide a temp principal, which should be fine since it is only persisted after the canister is created.